lib/src/main/java/graphql/nadel/engine/transform/partition/NadelPartitionMutationPayloadMerger.kt

```diff
@@ -0,0 +1,87 @@
+package graphql.nadel.engine.transform.partition
+
+import graphql.nadel.engine.transform.result.NadelResultInstruction
+import graphql.nadel.engine.transform.result.json.JsonNode
+import graphql.nadel.engine.util.unwrapNonNull
+import graphql.normalized.ExecutableNormalizedField
+import graphql.schema.GraphQLList
+import graphql.schema.GraphQLObjectType
+import graphql.schema.GraphQLOutputType
+import graphql.schema.GraphQLScalarType
+
+object NadelPartitionMutationPayloadMerger {
+
+    fun mergeDataFromMutationPayloadLike(
+        dataFromPartitionCalls: List<Any?>,
+        thisNodesData: Any?,
+        parentNodes: List<JsonNode>,
+        overallField: ExecutableNormalizedField,
+    ): List<NadelResultInstruction.Set> {
+        val parentNode = parentNodes.first()
+
+        // TODO: I'm not sure what's the best way to handle non-successful calls
+        // - force `success` to be false?
+        // - leave `success` alone?
+        val nonSuccessPayload = mapOf("success" to false)
+
+        val mutationPayloadLikeDataFromPartitionCalls = dataFromPartitionCalls
+            .mapNotNull {
+                if (it == null) {
+                    nonSuccessPayload
+                } else {
+                    check(it is Map<*, *>) { "Expected a Map, but got ${it::class.simpleName}" }
+                    it
+                }
+            }
+
+        val thisNodesDataCast = thisNodesData?.let {
+            check(it is Map<*, *>) { "Expected a Map, but got ${it::class.simpleName}" }
+            it
+        } ?: nonSuccessPayload
+
+        val allListKeys = (thisNodesDataCast.keys + mutationPayloadLikeDataFromPartitionCalls.flatMap { it.keys })
+            .distinct()
+
+        val mergedData = mutationPayloadLikeDataFromPartitionCalls.fold(thisNodesDataCast) { acc, next ->
+            allListKeys.associate {
+                if (it == "success") {
+                    Pair(it, acc[it].safeToBoolean() && next[it].safeToBoolean())
+                } else {
+                    if(acc[it] == null && next[it] == null) {
+                        Pair(it, null)
+                    } else {
+                        Pair(it, acc[it].safeToList() + next[it].safeToList())
+                    }
+                }
+            }
+        }
+
+        return listOf(
+            NadelResultInstruction.Set(
+                subject = parentNode,
+                newValue = JsonNode(mergedData),
+                field = overallField
+            )
+        )
+    }
+
+    /**
+     * A GraphQL type is considered to be a "mutation payload" (for the purposes of this transform)  if
+     * it has a `success` field of type `Boolean!` plus an `errors` field of type List and any other number
+     * of fields of type List
+     */
```

Comment:
I think atlassian public schema defines Payload as
```graphql
"""
The general shape of a mutation response.
"""
interface Payload {
    "Was this mutation successful"
    success: Boolean!

    "A list of errors if the mutation was not successful"
    errors: [MutationError!]
}
```
would it be reasonable to check if the type extends `Payload`

Reply:
I considered it, but eventually decided not to do it because the type for the use case we're targeting doesn't actually implement `Payload`. But you make a good point, that type should really be implementing this interface. They're already adhering to the contract, so it shouldn't be a big change for them 👍